### ISSUE TITLE
iOS Firebase storage upload works on first upload and then fails on second and subsequent uploads

### ISSUE THREAD

**Issue by analogjedi** [firebase/firebase-ios-sdk#6720]
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: _12.01____
  * Firebase SDK version: __6.34__and__6.31.1__(the 2 versions tried)_
  * Firebase Component: __Storage___ (Auth, Core, Database, Firestore, Messaging, Storage, etc)
  * Component version: __3.9.1__and__3.9.0___
  * Installation method: `CocoaPods 

### [REQUIRED] Step 2: Describe the problem
https://stackoverflow.com/questions/64306854/ios-swift-firebase-storage-upload-error-only-the-initial-file-upload-works

Problem: I initialize Firebase in iOS app and successfully write database data to Cloud FireStore. I then upload related video file to Firebase Storage with an asynch call. The first upload works with the first recorded video. Another video is recorded and the second or further attempts to upload the .mov file fail

cloud storage VIDEO file upload error: Error Domain=FIRStorageErrorDomain Code=-13000 "An unknown error occurred, please check the server response." UserInfo={object=PBY7Ost7nPWD8jWWF4qG.mov, ResponseBody=Can not finalize upload. Current size is 1692167. Expected final size is 1665242., bucket=launch-me-47860.appspot.com, data={length = 83, bytes = 0x43616e20 6e6f7420 66696e61 6c697a65 ... 31363635 3234322e }, data_content_type=text/plain; charset=utf-8, NSLocalizedDescription=An unknown error occurred, please check the server response., ResponseErrorDomain=com.google.HTTPStatus, ResponseErrorCode=400}

#### Steps to reproduce:

Cloud Firestore database writes ALWAYS work.
Using the index generated from Cloud Firestore write, Firebase storage is then used to upload a video and a data file with names as the index from (1) above.
The first first file upload always works with a new app launch. The second or any additional file upload fails with the following error:

I have found similar questions posted from 2016 and 2017 but this Firebase write problem seems different as the initial upload always works and then the next time I try to perform an upload it fails with the error.

#### Relevant Code:

```
// MARK: Write file to Firebase Cloud Storage
    private func fbCloudFileWrite(indexName: String) {
        let fbStorage = Storage.storage()
        print("(DEBUG FB) fbCloudFileWrite: upload indexName.csv and indexName.mov")
        // now upload file to cloud FireStore
        let fbStorageRef = fbStorage.reference()
        // Create a reference to the file you want to upload
        //let LaunchMeDataRef = fbStorageRef.child("LaunchMe/" + indexName + ".csv")
        //let LaunchMeVideoRef = fbStorageRef.child("LaunchMe/" + indexName + ".mov")
        var LaunchMeDataRef = fbStorageRef.child(indexName + ".csv")
        var LaunchMeVideoRef = fbStorageRef.child(indexName + ".mov")
        // Upload the file to the path "images/rivers.jpg"
        fbWriteAttempts += 1
        DispatchQueue.main.async {
            print("(DEBUG FB) ***** write attemp #: \(self.fbWriteAttempts)")
            if let vURL = self.videoURL, let dURL = self.dataFileURL {
                let uploadVideoTask = LaunchMeVideoRef.putFile(from: vURL, metadata: nil) { metadata, err in
                    if let err = err {
                        print("(DEBUG FB) cloud storage VIDEO file upload error: \(err)")
                        if self.fbWriteAttempts < 4 {
                            self.fbCloudFileWrite(indexName: indexName)
                        }
                    } else {
                        print("(DEBUG FB) video uploaded: \(indexName)")
                        let uploadDataTask = LaunchMeDataRef.putFile(from: dURL, metadata: nil) { metadata, err in
                            if let err = err {
                                print("(DEBUG FB) cloud storage SENSOR file upload error: \(err)")
                            } else {
                                print("(DEBUG FB) sensor data uploaded: \(indexName)")
                                print("(DEBUG FB) set newRecording = false to prevent duplicates")
                                // all files successfully uploaded. Set newRecording to false
                                self.newRecording = false
                            }
                        } // close uploadDataTask
                    } // close else
                } // close let uploadVideoTask
            } // close vURL unwrap
            else {
                print("(DEBUG FB) videoURL could not be unwrapped")
            }
        } // close Dispatch.main.async
```

**Comment by analogjedi** [comment]
I was able to resolve this issue by setting the URL in the viewController that saves the file instead of passing the URL from the previous viewController:

private func setVideoStorageURL() {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        let path = "output.mov"
        videoURL = paths[0].appendingPathComponent(path)
    }

It is very strange since the file size is printed properly to the log file prior to Firebase Storage write yet the file upload reports that the expected file size is actually the file size from the previous video recording.

**Comment by schmidt-sebastian** [comment]
@analogjedi Thanks for filing this. I tried to reproduce this, but wasn't able to get my test case to fail. Do you see anything that is meaningfully different in my reproduction?

```
func doubleUpload() throws {
    let expectation = self.expectation(description: #function)
    
    let tmpDir = URL(fileURLWithPath: NSTemporaryDirectory())
    
    let ref1 = storage.reference(withPath: "ios/public/testSimplePutFile1")
    let data1 = "Hello".data(using: .utf8)
    let fileURL1 = tmpDir.appendingPathComponent("hello1.txt")
    try data1?.write(to: fileURL1, options: .atomicWrite)
    
    let ref2 = storage.reference(withPath: "ios/public/testSimplePutFile2")
    let data2 = "Hello World".data(using: .utf8)
    let fileURL2 = tmpDir.appendingPathComponent("hello2.txt")
    try data2?.write(to: fileURL2, options: .atomicWrite)
    
    DispatchQueue.main.async {
      ref1.putFile(from: fileURL1, metadata: nil) { metadata, error in
         XCTAssertNil(error, "Error should be nil")
         ref2.putFile(from: fileURL2, metadata: nil) { metadata, error in
            XCTAssertNil(error, "Error should be nil")
            expectation.fulfill()
          }
       }
    }
    waitForExpectations()
 }
```

I also looked through the code and couldn't see anything obvious. The size is computed just once before the upload, but as long as the file doesn't get modified during the upload, I don't see how this would corrupt the data.

**Comment by analogjedi** [comment]
@schmidt-sebastian , thank you for the reply. I uploaded the stack overflow question with my answer:
[https://stackoverflow.com/questions/64306854/ios-swift-firebase-storage-upload-error-only-the-initial-file-upload-works](https://stackoverflow.com/questions/64306854/ios-swift-firebase-storage-upload-error-only-the-initial-file-upload-works)

I do not see anything meaningfully different except it was only a problem with my movie file.
I would create a video in one controller, then pass the URL to the second controller.
The second controller would playback the correct video which implies the correct URL yet the Firebase Storage upload, when provided the same URL, would complain of expected size difference and the size it expected would be from the previous video. It is quite strange.

When I go from one controller to the next without recording a new video then the upload always works because the size would match the previous video. When I force a new URL set with the added function in the controller that performs the Firebase Upload then it works. This is what I posted as my answer on stack overflow. 

A mystery that is perhaps not worth solving. Thank you for your support!
